### DISCUSSION TITLE
Changing show rules in multiple levels of headings

### DISCUSSION THREAD

**Discussion by psads-git** [typst/typst#3301] [Q&A] [answered]
I am trying to change the `show` rules of levels 2 and 3 of the headings, to remove the `numbering`. I am using the code below, which works when I uncomment the commented part. However, the compacted way that uses a `for` loop does not work. Could you please explain why the `for` loop approach does not work? Thanks!

```
#set heading(numbering: "1.")

#for i in (2, 3) {
  show heading.where(
    level: i
  ): it => {
    [#strong(it.body)]
  }
}

// #show heading.where(
//     level: 2
//   ): it => {
//     [#strong(it.body)]
//   }

// #show heading.where(
//     level: 3
//   ): it => {
//     [#strong(it.body)]
//   }


= head 1

== head 2

=== head 3

```

**Comment by xkevio** [comment]
Because in the first case, the `show`-rule is scoped to only affect everything inside of the `for`-loop.
In the second case, their scope is global.

**Reply by psads-git** [reply]
Thanks, @xkevio, for your clarification!

So, there is no more compact way of setting a show rule for multiple levels than setting a show rule for each level, without avoiding code repetition?

**Reply by laurmaedje** [reply, marked as answer]
It is possible like this:
```typ
#set heading(numbering: "1.")
#show: rest => {
  for i in (2, 3) {
    rest = {
      show heading.where(level: i): it => strong(it.body)
      rest
    }
  }
  rest
}

= head 1
== head 2
=== head 3
```

<details>
  <summary>
  There are plans to make programmatic manipulation of show rules like this simpler, along these lines (but it's not yet implemented):
  </summary>

  ```typ
  #apply for i in (2, 3) {
    show heading.where(level: i): it => strong(it.body)
  }
  ```
  Here the show rules would join together like content does, but an explicit `apply` would be necessary to ensure that the show rules "leak" intentionally.

  What would also help (but is also unimplemented) is support for "or"-selectors in show rules (these currently only work with `query()`). Then, you could write something like:
```typ
#show heading.where(level: 2).or(heading.where(level: 3)): ..
```
</details>

**Reply by Julian-Wassmann** [reply]
Another option is to reconstruct the heading in a show rule. Note the `parbreak()`, which is important if you want the heading in its own line.
```typst
#show heading: it => {
  if it.numbering != none and not it.level in (2, 3) {
    counter(heading).display()
    h(0.3em)
  }
  it.body
  parbreak()
}
```

**Reply by psads-git** [reply]
Thanks, @laurmaedje, for your solution and explanations! The improvements you are planning appear to be in the right direction!

**Reply by psads-git** [reply]
Thanks, @Julian-Wassmann, for your solution!
